prow/servemux/servemux.go

```diff
@@ -0,0 +1,399 @@
+}
+
+func (o *ServeMuxHandlerOptions) RegisterIssueEventHandler(fn issueEventHandler) {
+	o.serveMuxHandler.issueEventHandlers = append(o.serveMuxHandler.issueEventHandlers, fn)
+}
+
+func (o *ServeMuxHandlerOptions) RegisterStatusEventHandler(fn statusEventHandler) {
+	o.serveMuxHandler.statusEventHandlers = append(o.serveMuxHandler.statusEventHandlers, fn)
+}
+
+func (o *ServeMuxHandlerOptions) RegisterExternalPlugins(p map[string][]plugins.ExternalPlugin) {
+	o.serveMuxHandler.externalPlugins = p
+}
+
+func (o *ServeMuxHandlerOptions) RegisterHelpProvider(helpProvider func([]config.OrgRepo) (*pluginhelp.PluginHelp, error), log *logrus.Entry) {
+	pluginhelp_externalplugins.ServeExternalPluginHelp(o.httpServeMux, log, helpProvider)
+}
+
+func (o *ServeMuxHandlerOptions) RegisterPluginHelpAgentHandle(endpoint string, helpAgent *pluginhelp_hook.HelpAgent) {
+	o.httpServeMux.Handle(endpoint, helpAgent)
+}
+
+func (o *ServeMuxHandlerOptions) RegisterCustomFuncHandle(endpoint string, fn func(w http.ResponseWriter, r *http.Request)) {
+	o.httpServeMux.HandleFunc(endpoint, fn)
+}
+
+// GracefulShutdown handles all requests sent before receiving the shutdown signal.
+func (o *ServeMuxHandlerOptions) GracefulShutdown() {
+	logrus.Info("Waiting for the remaining requests")
+	o.wg.Wait()
+}
+
+// ServeMuxHandler is a http serveMux handler that implements the ServeHTTP method.
+// It can be used in any external plugin to serve an HTTP server with this handler.
+// see https://godoc.org/net/http#ServeMux
+type serveMuxHandler struct {
+	log *logrus.Entry
+	wg  *sync.WaitGroup
+
+	reviewCommentEventHandlers []reviewCommentEventHandler
+	reviewEventHandlers        []reviewEventHandler
+	pullRequestHandlers        []pullRequestHandler
+	pushEventHandlers          []pushEventHandler
+	issueCommentEventHandlers  []issueCommentEventHandler
+	issueEventHandlers         []issueEventHandler
+	statusEventHandlers        []statusEventHandler
+
+	externalPlugins map[string][]plugins.ExternalPlugin
+
+	hmacTokenGenerator func() []byte
+	metrics            *Metrics
+
+	c http.Client
+}
+
+func (s *serveMuxHandler) handleEvent(eventType, eventGUID string, payload []byte, h http.Header) error {
+	var org string
+	var repo string
+
+	l := logrus.WithFields(logrus.Fields{eventTypeField: eventType, github.EventGUID: eventGUID})
+
+	// We don't want to fail the webhook due to a metrics error.
+	if counter, err := s.metrics.WebhookCounter.GetMetricWithLabelValues(eventType); err != nil {
+		l.WithError(err).Warn("Failed to get metric for eventType " + eventType)
+	} else {
+		counter.Inc()
+	}
+
+	switch eventType {
+	case "issues":
+		var i github.IssueEvent
+		if err := json.Unmarshal(payload, &i); err != nil {
+			return err
+		}
+		i.GUID = eventGUID
+		org = i.Repo.Owner.Login
+		repo = i.Repo.Name
+
+		for _, issueEventHandler := range s.issueEventHandlers {
+			s.wg.Add(1)
+			go issueEventHandler(l, i, s.wg)
```

Comment:
Instead of passing in the workgroup everywhere and relying on consumers to correctly call `wg.Done`, why do we not simply wrap it:
```
go func(){
  defer wg.Done()
  issueEventhandler(l, i)
}
```

Reply:
that sounds better :)

Reply:
However, on a second look, the consumers will spawn different goroutines and in case of interruption, we will kill all of those.

Reply:
Where do they spawn different goroutines? They just register a plugin for an event type and we invoke that plugin for the given event, the plugin itself is not supposed to start new goroutines? And if it does it should just have its own waitGroup and not return until that is done?

Reply:
You are right. I think its better to let the consumer handles its own waitgroups. Let me address this.